src/provider/provider_tracking.c

```diff
@@ -119,6 +121,159 @@ static umf_result_t trackingAlloc(void *hProvider, size_t size,
+        // TODO: what now? should we rollback the split? This is can only happen due to ENOMEM
+        // so it's unlikely but probably the best solution would be to try to preallocate everything
+        // (value and critnib nodes) before calling umfMemoryProviderAllocationSplit.
+        goto err;
+    }
+
+    int cret = critnib_insert(p->hTracker->map, (uintptr_t)ptr,
+                              (void *)splitValue, 1 /* update */);
+    // this cannot fail since we know the element exists (nothing to allocate)
+    assert(cret == 0);
+    (void)cret;
+
+    // free the original value
+    free(value);
+    util_mutex_unlock(p->hTracker->splitMergeMutex);
+
+    return UMF_RESULT_SUCCESS;
+
+err:
+    util_mutex_unlock(p->hTracker->splitMergeMutex);
+err_lock:
+    free(splitValue);
+    return ret;
+}
+
+static umf_result_t trackingAllocationMerge(void *provider, void *lowPtr,
+                                            void *highPtr, size_t totalSize) {
+    umf_result_t ret = UMF_RESULT_ERROR_UNKNOWN;
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+
+    tracker_value_t *mergedValue =
+        (tracker_value_t *)malloc(sizeof(tracker_value_t));
+    mergedValue->pool = p->pool;
+    mergedValue->size = totalSize;
+
+    if (!mergedValue) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    int r = util_mutex_lock(p->hTracker->splitMergeMutex);
+    if (r) {
+        goto err_lock;
+    }
+
+    tracker_value_t *leftValue =
+        (tracker_value_t *)critnib_get(p->hTracker->map, (uintptr_t)lowPtr);
+    if (!leftValue) {
+        fprintf(stderr, "tracking merge: no left value\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+    tracker_value_t *rightValue =
+        (tracker_value_t *)critnib_get(p->hTracker->map, (uintptr_t)highPtr);
+    if (!rightValue) {
+        fprintf(stderr, "tracking merge: no right value\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+    if (leftValue->pool != rightValue->pool) {
+        fprintf(stderr, "tracking merge: pool mismatch\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+    if (leftValue->size + rightValue->size != totalSize) {
+        fprintf(stderr, "tracking merge: leftValue->size + rightValue->size != "
+                        "totalSize\n");
+        ret = UMF_RESULT_ERROR_INVALID_ARGUMENT;
+        goto err;
+    }
+
+    ret = umfMemoryProviderAllocationMerge(p->hUpstream, lowPtr, highPtr,
+                                           totalSize);
+    if (ret != UMF_RESULT_SUCCESS) {
+        fprintf(stderr,
+                "tracking merge: umfMemoryProviderAllocationMerge failed\n");
+        goto err;
+    }
+
+    // We'll have duplicate entry for the range [highPtr, rightValue->size] but this is fine,
+    // the value is the same anyway and we forbid splitting/removing that range concurrently
```

Comment:
splitting -> merging ?

Reply:
Just removing now, since split/merge is called under the lock.